compiler/lib/source_map.ml

```diff
@@ -438,8 +438,116 @@ let of_json (json : Yojson.Raw.t) =
       }
   | _ -> invalid ()
 
-let of_string s = of_json (Yojson.Raw.from_string s)
-
 let to_string m = Yojson.Raw.to_string (json m)
 
 let to_file m file = Yojson.Raw.to_file file (json m)
+
+module Index = struct
+  type offset =
+    { gen_line : int
+    ; gen_column : int
+    }
+
+  (* Type synonym to avoid confusion between toplevel [t] and this submodule's [t]. *)
+  type map = t
+
+  type t =
+    { version : int
+    ; file : string
+    ; sections : (offset * [ `Map of map ]) list
+    }
+
+  let json t =
+    let stringlit s = `Stringlit (Yojson.Safe.to_string (`String s)) in
+    `Assoc
+      [ "version", `Intlit (string_of_int t.version)
+      ; "file", stringlit (rewrite_path t.file)
+      ; ( "sections"
+        , `List
+            (List.map
+               ~f:(fun ({ gen_line; gen_column }, `Map sm) ->
+                 `Assoc
+                   [ ( "offset"
+                     , `Assoc
+                         [ "line", `Intlit (string_of_int gen_line)
+                         ; "column", `Intlit (string_of_int gen_column)
+                         ] )
+                   ; "map", json sm
+                   ])
+               t.sections) )
+      ]
+
+  let intlit ~errmsg name json =
+    match List.assoc name json with
+    | `Intlit i -> int_of_string i
+    | _ -> invalid_arg errmsg
+    | exception Not_found -> invalid_arg errmsg
+
+  let section_of_json : Yojson.Raw.t -> offset * [ `Map of map ] = function
+    | `Assoc json ->
+        let offset =
+          match List.assoc "offset" json with
+          | `Assoc fields ->
+              let gen_line =
+                intlit
+                  "line"
+                  fields
+                  ~errmsg:
+                    "Source_map_io.Index.of_json: field 'line' absent or invalid from \
+                     section"
+              in
+              let gen_column =
+                intlit
+                  "column"
+                  fields
+                  ~errmsg:
+                    "Source_map_io.Index.of_json: field 'column' absent or invalid from \
+                     section"
+              in
+              { gen_line; gen_column }
+          | _ ->
+              invalid_arg "Source_map_io.Index.of_json: 'offset' field of unexpected type"
+        in
+        (match List.assoc "url" json with
+        | _ ->
+            invalid_arg
+              "Source_map_io.Index.of_json: URLs in index maps are not currently \
```

Comment:
Is this a limitation of the spec or of this implementation ( why dont we support it ?)

Reply:
I found two sources that disagree on the presence of the url form in the spec.

Reply:
It's part of [the spec](https://sourcemaps.info/spec.html), but since the Chrome DevTools does not support it, I'm not sure many tools generate such files.
https://github.com/ChromeDevTools/devtools-frontend/blob/d8095064b341d27f160550b05904a86d05b7ac82/front_end/core/sdk/SourceMap.ts#L211-L216

Reply:
And we happen not to need it.

Reply:
https://tc39.es/source-map/#section-object doesn't mention it

Reply:
Apparently, no one supports this, so it has been dropped from the spec. See https://github.com/tc39/source-map/issues/29.